File: crate/src/cell/cell.rs
use std::fmt::{self, Display};
use crate::cell::{SetMethod, SYMBOLS, OptionsRemaining};
use crate::utils::bit_utils::{highest_bit_position, number_of_bits_set, power_of_2_bit_positions};

#[derive(Debug, Copy, Clone)]
pub struct Cell {
  max_cells: usize,
  max_columns: usize,

  pub column: usize,
  pub row: usize,
  pub options: u64,
  pub total_options_remaining: usize,

  pub set_method: SetMethod
}

impl Display for Cell {
  fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
    if formatter.alternate() {
      write!(formatter, "{:0>1$b}", self.options, (self.max_cells + 1) * self.max_columns - 1)
    } else {
      write!(formatter, "{:>3}", self.options)
    }
  }
}

impl Cell {
  pub fn new(max_columns: usize, max_rows: usize, column: usize, row: usize) -> Self {
    let max_cells = max_columns * max_rows;
    Cell {
      max_cells,
      max_columns,
      column,
      row,
      options: (1 << max_cells) - 1,                                // Set all bits
      total_options_remaining: max_cells,
      set_method: SetMethod::Unset
    }
  }

  pub fn reset(&mut self) {
    self.set_method = SetMethod::Unset;
    self.total_options_remaining = self.max_cells;
    self.options = (1 << self.max_cells) - 1; // Set all bits
  }

  pub fn equal(&self, cell: &Cell) -> bool {
    self.options == cell.options
  }

  pub fn solved(&self) -> bool {
    self.total_options_remaining == 1
  }

  pub fn symbol(&self) -> char {
    let index = highest_bit_position(self.options);
    let set_column = index % self.max_columns;
    let set_row = index / self.max_columns >> 0;

    SYMBOLS[set_row * self.max_columns + set_column]
  }

  pub fn remove_option_at_position(&mut self, column: usize, row: usize) -> bool {
    // Return if last option left after removing this option
    let mut last_option_found = false;
    let bit = 1 << self.max_columns * row + column;

    if self.options & bit > 0 {
      // Check if option to remove exists
      self.options &= !bit;
      self.total_options_remaining -= 1;
      if self.total_options_remaining == 1 {
        self.set_method = SetMethod::Calculated;
        last_option_found = true;
      }
    }

    last_option_found
  }

  pub fn remove_option(&mut self, option: u64) -> bool {
    // Return if last option left after removing this option
    let mut last_option_found = false;

    if self.options & option > 0 && self.total_options_remaining > 1 {
      // Check if option to remove exists and not last option
      self.options &= !option;
      self.total_options_remaining -= 1;
      if self.total_options_remaining == 1 {
        self.set_method = SetMethod::Calculated;
        last_option_found = true;
      }
    }

    last_option_found
  }

  pub fn remove_options(&mut self, remove: u64) -> bool {
    let mut last_option_found = false;

    let mut remove_options = self.options & remove;
    if remove_options > 0 && self.total_options_remaining > 1 && self.options & !remove > 0 {
      // Remove options iff cell contains other options
      self.options -= remove_options;
      self.total_options_remaining -= number_of_bits_set(remove_options);

      if self.total_options_remaining == 1 {
        self.set_method = SetMethod::Calculated;
        last_option_found = true;
      } else {
        while remove_options > 0 {
          let highest_bit_pos = highest_bit_position(remove_options);
          remove_options -= 1 << highest_bit_pos;
        }
      }
    }

    last_option_found
  }

  pub fn set_by_position(&mut self, column: usize, row: usize, set_method: SetMethod) {
    self.set_method = set_method;
    self.clear_all_except_at_position(column, row);
  }

  pub fn set_by_index(&mut self, index: usize, set_method: SetMethod) {
    self.set_method = set_method;
    self.clear_all_except_at_position(index % self.max_columns, index / self.max_columns >> 0);
  }

  pub fn set_by_option(&mut self, option: u64, set_method: SetMethod) {
    self.set_by_index(power_of_2_bit_positions(option), set_method);
  }

  pub fn set_by_symbol(&mut self, symbol: char, set_method: SetMethod) {
    self.set_by_index(find_symbol_index(symbol), set_method);
  }

  pub fn contains_option(&self, option: u64) -> bool {
    self.options & option > 0
  }

  pub fn contains_option_at_position(&self, column: usize, row: usize) -> bool {
    let bit = 1 << row * self.max_columns + column;
    (self.options & bit) > 0
  }

  pub fn contains_options(&self, check_options: u64) -> bool {
    self.options & check_options == check_options
  }

  pub fn contains_symbol(&self, symbol: char) -> bool {
    self.options & 1 << find_symbol_index(symbol) > 0
  }

  fn clear_all_except_at_position(&mut self, column: usize, row: usize) {
    self.options = 1 << self.max_columns * row + column;
    self.total_options_remaining = 1;
  }

  pub fn removed_options_per_row(&mut self, row: usize) -> Vec<usize> {
    let mut removed_options = Vec::with_capacity(self.max_columns);

    let mut bit = 1 << row * self.max_columns;
    for column in 0..self.max_columns {
      if self.options & bit == 0 {
        removed_options.push(column);
      }
      bit <<= 1;                                                    // bit = 1 << row * columns + column
    }

    removed_options
  }
}

fn find_symbol_index(symbol: char) -> usize {
  SYMBOLS.iter().position(|&x| x == symbol).unwrap()
}

File: src/solver/cell.ts
import {powerOf2BitPositions, highestBitPosition, numberOfBitsSet} from "./utils/bitUtilities";

export enum SetMethod {
	loaded = 'loaded',
	user = 'user',
	calculated = 'calculated'
}

export interface IJsonCellColumn {
	symbol?: string;
	strikeOut?: boolean;
	highlight?: boolean;
}

export interface IJsonCellRow {
	columns: IJsonCellColumn[];
}

export interface IJsonCell {
	rows?: IJsonCellRow[];
	symbol?: string;
	setMethod?: SetMethod;
}

export interface ICell {
	setMethod: SetMethod;
	readonly options: number;
	json: IJsonCell;
	totalOptionsRemaining: number;

	equal(cell: ICell): boolean;

	symbol(): string;
	getColumn(): number;
	getRow(): number;
	solved(): boolean;
	removeOption(option: number): boolean;
	removeOptionAtPosition(column: number, row: number): boolean;
	toggleRemoveOptionAtPositionShallow(column: number, row: number): void;
	toggleHighlightOptionAtPosition(column: number, row: number): void;
	removeOptions(remove: number): boolean;
	setByPosition(column: number, row: number, setMethod: SetMethod): void;
	setByOption(option: number, setMethod: SetMethod): void;
	setBySymbol(symbol: string, setMethod: SetMethod): void;
	setByIndex(index: number, setMethod: SetMethod): void;
	reset(): void;
	containsOption(option: number): boolean;
	containsOptionAtPosition(column: number, row: number): boolean;
	containsOptions(checkOptions: number): boolean;
	containsSymbol(symbol: string): boolean;
	removedOptionsPerRow(row: number): number[];
	setJson(json: IJsonCell): void;

	symbolPosition(): number;
	setOption(): number;
}

export class Cell implements ICell {
	private static symbols: string;
	private static columns: number;
	private static rows: number;

	public setMethod: SetMethod;
	public options: number;
	public setColumn: number;
	public setRow: number;
	public json: IJsonCell;

	public totalOptionsRemaining: number;

	static Constructor(columns: number, rows: number) {               // Static constructor
		Cell.symbols = "123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ0"
		Cell.columns = columns;
		Cell.rows = rows;
	}

	// constructor(column: number, row: number);
	// constructor(copy: ICell);
	constructor(public column: any, public row?: number) {
		if (typeof column === "number") {
			this.reset();
		} else {                                                     		// Copy constructor
			const copy: Cell = column;
			this.column = copy.column;
			this.row = copy.row;
			this.setMethod = copy.setMethod;

			this.options = copy.options;
			this.setColumn = copy.setColumn;
			this.setRow = copy.setRow;
			this.totalOptionsRemaining = copy.totalOptionsRemaining;

			this.json = copy.json;
			
			if (copy.json.rows) {
				this.json = { rows: [] };
				for (let row: number = 0; row < copy.json.rows.length; row++) {
					this.json.rows[row] = { columns: [] };
					const jsonColumns: IJsonCellColumn[] = copy.json.rows[row].columns;
					for (let column = 0; column < jsonColumns.length; column++) {
						this.json.rows[row].columns[column] = { symbol: jsonColumns[column].symbol };
						if (jsonColumns[column].strikeOut) {
							this.json.rows[row].columns[column].strikeOut = jsonColumns[column].strikeOut;
						}
					}
				}
			} else {
				this.json = { symbol: copy.json.symbol, setMethod: copy.json.setMethod };
			} 
		}
	}

	public reset() {
		this.setMethod = null;

		this.options = (1 << Cell.columns * Cell.rows) - 1;            	// Set all bits
		this.setColumn = -1;
		this.setRow = -1;
		this.totalOptionsRemaining = Cell.columns * Cell.rows;

		this.json = { rows: [] };                                  			// Set JSON representation to all options available 
		for (let row: number = 0, index: number = 0; row < Cell.rows; row++) {
			const columns: IJsonCellColumn[] = [];
			for (let column = 0; column < Cell.columns; column++) {
				columns.push({ symbol: Cell.symbols[index++] });
			}
			this.json.rows.push({ columns: columns });
		}
	}

	public equal(cell: Cell): boolean {
		return (this.setColumn === cell.setColumn || cell.setColumn === -1) &&
			(this.setRow === cell.setRow || cell.setRow === -1) && this.options === cell.options;
		//return this.options === cell.options;
	}

	public symbol(): string {
		return Cell.symbols[this.setRow * Cell.columns + this.setColumn];
	}

	public getColumn(): number {
		return this.column;
	}

	public getRow(): number {
		return this.row;
	}

	public solved(): boolean {
		return this.totalOptionsRemaining === 1;
	}

	public removeOptionAtPosition(column: number, row: number): boolean {	// Return if last option left after removing this option
		let lastOptionFound: boolean = false;

		const bit: number = 1 << Cell.columns * row + column;
		if (this.options & bit) {                                				// Check if option to remove exists
			this.options &= ~bit;
			if (--this.totalOptionsRemaining === 1) {
				this.setRemainingOption(this.options);                     	// Set last remaining option's column and row 
				this.json = { symbol: Cell.symbols[powerOf2BitPositions[this.options]] };
				this.setMethod = SetMethod.calculated;
				lastOptionFound = true;
			} else {
				this.json.rows[row].columns[column].strikeOut = true; 			// Only set strikeOut to true if option removed - else leave empty   
			}
		}

		return lastOptionFound;
	}

	public toggleRemoveOptionAtPositionShallow(column: number, row: number): void {
		const cell = this.json.rows[row].columns[column];
		cell.strikeOut = !cell.strikeOut;
	}

	public toggleHighlightOptionAtPosition(column: number, row: number): void {
		const cell = this.json.rows[row].columns[column];
		cell.highlight = !cell.highlight;
	}

	public removeOption(option: number): boolean {                   	// Return if last option left after removing this option
		let lastOptionFound = false;
		// debugger

		console.log!(`=====>>> (${this.column}, ${this.row}) (${this.setColumn}, ${this.setRow}) ${this.options} ${option} ${this.totalOptionsRemaining}`);

		if (this.options & option && this.totalOptionsRemaining > 1) { 	// Check if option to remove exists and not last option
			this.options &= ~option;
			if (--this.totalOptionsRemaining === 1) {
				this.setRemainingOption(this.options);                     	// Set last remaining option's column and row 
				this.json = { symbol: Cell.symbols[powerOf2BitPositions[this.options]] };
				this.setMethod = SetMethod.calculated;
				lastOptionFound = true;
			}
			else {
				const index: number = powerOf2BitPositions[option];
				this.json.rows[index / Cell.columns >> 0].columns[index % Cell.columns].strikeOut = true;
			}
		}

		return lastOptionFound;
	}

	public removeOptions(remove: number): boolean {
		let lastOptionFound = false;

		let removeOptions = this.options & remove;
		if (removeOptions && this.totalOptionsRemaining > 1 && this.options & ~remove) {	// Remove options iff cell contains other options
			this.options -= removeOptions;
			this.totalOptionsRemaining -= numberOfBitsSet(removeOptions);

			if (this.totalOptionsRemaining === 1) {
				this.setRemainingOption(this.options);                			// Set last remaining option's column and row 
				this.json = { symbol: Cell.symbols[powerOf2BitPositions[this.options]] };
				this.setMethod = SetMethod.calculated;
				lastOptionFound = true;
			}
			else {
				while (removeOptions) {
					let highestBitPos = highestBitPosition(removeOptions);
					this.json.rows[highestBitPos / Cell.columns >> 0].columns[highestBitPos % Cell.columns].strikeOut = true;
					removeOptions -= 1 << highestBitPos;
				}
			}
		}

		return lastOptionFound;
	}

	public setByPosition(column: number, row: number, setMethod: SetMethod) {
		this.clearAllExceptAtPosition(this.setColumn = column, this.setRow = row, this.setMethod = setMethod);
	}

	public setByIndex(index: number, setMethod: SetMethod) {
		this.clearAllExceptAtPosition(this.setColumn = index % Cell.columns, this.setRow = index / Cell.columns >> 0, this.setMethod = setMethod);
	}

	public setByOption(option: number, setMethod: SetMethod) {
		// console.log(option)
		this.setByIndex(powerOf2BitPositions[option], setMethod);
	}

	public setBySymbol(symbol: string, setMethod: SetMethod){
		this.setByIndex(Cell.symbols.indexOf(symbol), setMethod);
	}

	public containsOption(option: number): boolean {
		return (this.options & option) > 0;
	}

	public containsOptionAtPosition(column: number, row: number): boolean {
		const bit: number = 1 << row * Cell.columns + column;
		return (this.options & bit) > 0;
	}

	public containsOptions(checkOptions: number): boolean {
		return (this.options & checkOptions) === checkOptions;
	}

	public containsSymbol(symbol: string): boolean {
		const index = Cell.symbols.indexOf(symbol);
		return (this.options & 1 << index) > 0; // eslint-disable-line no-mixed-operators
	}

	private setRemainingOption(options: number) {
		const index = highestBitPosition(options);
		this.setColumn = index % Cell.columns;
		this.setRow = index / Cell.columns >> 0;
	}

	// private clearAllExcept(option: number, fix: boolean) {
	// 	this.options = option;
	// 	this.json = { symbol: Cell.symbols[powerOf2BitPositions[option]], fixed: fix };
	// 	this.totalOptionsRemaining = 1;
	// }

	private clearAllExceptAtPosition(column: number, row: number, setMethod: SetMethod) {
		this.options = 1 << Cell.columns * row + column;
		this.json = { symbol: Cell.symbols[powerOf2BitPositions[this.options]], setMethod };
		this.totalOptionsRemaining = 1;
	}

	public removedOptionsPerRow(row: number): number[] {
		const removedOptions: number[] = [];
		for (let column: number = 0, bit: number = 1 << row * Cell.columns; column < Cell.columns; column++, bit <<= 1) {	// bit = 1 << row * columns + column
			if (!(this.options & bit)) {
				removedOptions.push(column);
			}
		}

		return removedOptions;
	}

	// Remove options iff cell contains other options
	public removeIfExtraOptions(options: number): boolean {
		return this.totalOptionsRemaining > 1 && (this.options & ~options) > 0 && this.removeOptions(options);
	}

	public setJson(json: IJsonCell) {
		if (json.rows) {
			this.options = 0;
			this.totalOptionsRemaining = 0;
			for (let row: number = 0, option: number = 1; row < json.rows.length; row++) {
				const columns: IJsonCellColumn[] = json.rows[row].columns;
				for (let column = 0; column < columns.length; column++, option <<= 1) {
					if (!columns[column].strikeOut) {
						this.options += option;
						this.totalOptionsRemaining++;
					}
				}
			}
		} else {
			this.setBySymbol(json.symbol, json.setMethod);
		}

		this.json = json;
	}

	public symbolPosition() {
		return this.setMethod ? this.setRow * Cell.columns + this.setColumn + 1: 0;
	}

	public setOption() {
		return this.setMethod ? this.options: 0;
	}
}
